internal/util/logging/pgx.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !ferretdb_no_postgresql
+
+package logging
+
+import (
+	"context"
+	"log/slog"
+	"runtime"
+	"slices"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/jackc/pgx/v5/tracelog"
+	"golang.org/x/exp/maps"
+
+	"github.com/FerretDB/FerretDB/internal/util/debugbuild"
+	"github.com/FerretDB/FerretDB/internal/util/must"
+)
+
+// pgxLogLevels maps pgx log levels to slog levels.
+var pgxLogLevels = map[tracelog.LogLevel]slog.Level{
+	// pgx's info level is used for queries, but we want to hide them by default
+	tracelog.LogLevelTrace: slog.LevelDebug - 2,
+	tracelog.LogLevelDebug: slog.LevelDebug,
+	tracelog.LogLevelInfo:  slog.LevelDebug + 2,
+
+	tracelog.LogLevelWarn:  slog.LevelWarn,
+	tracelog.LogLevelError: slog.LevelError,
+}
+
+// pgxLogger is a pgx's [tracelog.Logger] implementation that uses slog
+// with correct source code location.
+type pgxLogger struct {
+	l *slog.Logger
+}
+
+// NewPgxLogger creates a new [tracelog.Logger] that uses given [slog.Logger].
+func NewPgxLogger(l *slog.Logger) tracelog.Logger {
+	return &pgxLogger{l: l}
+}
+
+// Log implements [tracelog.Logger].
+func (sl *pgxLogger) Log(ctx context.Context, level tracelog.LogLevel, msg string, data map[string]any) {
+	now := time.Now()
+
+	l, ok := pgxLogLevels[level]
+	if !ok {
+		l = slog.LevelError + 1
+		sl.l.LogAttrs(
+			ctx, slog.LevelWarn, "Invalid pgx log level",
+			slog.Int64("level", int64(level)), slog.String("original_msg", msg),
+		)
+	}
+
+	if !sl.l.Enabled(ctx, l) {
```

Comment:
Shouldn't this statement be first in the function?

Reply:
Why should it?

Reply:
Because if we are going to exit from this function, we don't need `now := time.Now()`.

Reply:
But it does not answer why you think that line should be the first in the function — there is some other code in between

Reply:
The code in between is even more confusing to me, I expected it to be a panic.

Reply:
Logger shouldn't panic, because often it's the last thing we use to debug. So panic in logging should be avoided at all cost.

The thing in the between converts `pgx` log level to `slog` log level. If pgx was configured to log level unknown to `slog`, we don't know how to map it. Best we can do is to log with a log level of our choice :hugs:

Reply:
Pgx documentation does not say what level values are used:
```go
// LogLevel represents the pgx logging level. See LogLevel* constants for
// possible values.
type LogLevel int

// The values for log levels are chosen such that the zero value means that no
// log level was specified.
const (
	LogLevelTrace = LogLevel(6)
	LogLevelDebug = LogLevel(5)
	LogLevelInfo  = LogLevel(4)
	LogLevelWarn  = LogLevel(3)
	LogLevelError = LogLevel(2)
	LogLevelNone  = LogLevel(1)
)
```

It is not clear if `LogLevelNone`, `LogLevel(0)`, and other values could be used.

On the other hand, that's what DPanic is for now. Updated.

`now := time.Now()` is on top to capture the time of the `Log` invocation as early as possible.